overlord/snapstate/component.go

```diff
@@ -170,9 +170,11 @@ func doInstallComponent(st *state.State, snapst *SnapState, compSetup *Component
 		addTask(kmodSetup)
 	}
 
+	changingSnapRev := !snapst.Current.Unset() && snapst.Current != snapSi.Revision
+
 	// We might be replacing a component
 	compInstalled := snapst.IsComponentInCurrentSeq(compSi.Component)
-	if compInstalled {
+	if compInstalled && !changingSnapRev {
```

Comment:
Are we sure we do not want to unlink the components? Would not make this things more difficult when we, for instance do a revert?

Reply:
I could see it going either way, but I don't really understand the connection to reverts? Since the links are per-snap-revision, we really only need to unlink if we're replacing something in that snap revision.

Reply:
I was thinking about "snap revert", where we would go back to the previous sequence point (previous snap revision), and would have to install again those components.

Reply:
That is a good point. Although in that case, we could be careful to not re-link during a revert, since the link should already be there. Either that, or we could make it idempotent. It seemed simpler to me to include/exclude tasks, rather than making the tasks handle many scenarios. For a bit of explanation of why I chose to not unlink:

In the unlink-current-component task, we `RemoveComponentForRevision` on the SnapState, which we don't want to do during a refresh, since the component should remain in the state for that revision (for things like reverts).

The complexity has to go somewhere, I guess. So I'm happy to do whichever way you think is best.

Reply:
Right, it is true that if we remove from the sequence then we would not know what was installed if reverting, so I guess that is fine. Also we keep mounts for snaps even after refrehing to a new version, so it would be coherent in that way. Maybe worth adding a comment with this information though.

Reply:
Please remember to drop a comment here